src/adapter/src/client.rs

```diff
@@ -482,52 +484,77 @@ impl SessionClient {
-                    | Command::Commit { .. }
-                    | Command::CancelRequest { .. }
-                    | Command::DumpCatalog { .. }
-                    | Command::CopyRows { .. }
-                    | Command::GetSystemVars { .. }
-                    | Command::SetSystemVars { .. }
-                    | Command::Terminate { .. } => {}
-                };
-                cmd
-            })
-            .await;
-        let status = if res.result.is_ok() {
-            "success"
-        } else {
-            "error"
-        };
-        if let Some(typ) = typ {
-            self.inner()
-                .inner
-                .metrics
-                .commands
-                .with_label_values(&[typ, status, name_hint.as_str()])
-                .inc();
+        let (tx, mut rx) = oneshot::channel();
+        let conn_id = session.conn_id();
+        let secret_key = session.secret_key();
+        self.inner_mut().inner.send({
+            let cmd = f(tx, session);
+            // Measure the success and error rate of certain commands:
+            // - declare reports success of SQL statement planning
+            // - execute reports success of dataflow execution
+            match cmd {
+                Command::Declare { .. } => typ = Some("declare"),
+                Command::Execute { .. } => typ = Some("execute"),
+                Command::Startup { .. }
+                | Command::Describe { .. }
+                | Command::VerifyPreparedStatement { .. }
+                | Command::Commit { .. }
+                | Command::CancelRequest { .. }
+                | Command::DumpCatalog { .. }
+                | Command::CopyRows { .. }
+                | Command::GetSystemVars { .. }
+                | Command::SetSystemVars { .. }
+                | Command::Terminate { .. } => {}
+            };
+            cmd
+        });
+
+        let mut cancel_future = pin::pin!(cancel_future);
+        let mut cancelled = false;
+        loop {
+            tokio::select! {
+                res = &mut rx => {
+                    let res = res.expect("sender dropped");
+                    let status = if res.result.is_ok() {
+                        "success"
+                    } else {
+                        "error"
+                    };
+                    if let Some(typ) = typ {
+                        self.inner()
+                            .inner
+                            .metrics
+                            .commands
+                            .with_label_values(&[typ, status, name_hint.as_str()])
+                            .inc();
+                    }
+                    self.session = Some(res.session);
+                    return res.result
+                },
+                _err = &mut cancel_future, if !cancelled => {
+                    cancelled = true;
+                    self.inner_mut().cancel_request(conn_id, secret_key);
```

Comment:
Do we need to maintain the `cancelled` flag? After calling `cancel_request(...)` can we eagerly return with an error instead of waiting for rx to close and take the above path? That should also allow us to eliminate the need for a `loop`?

Reply:
I think so yeah. cancel_request is asking controller to cancel but some things aren't cancellable (like an insert or an `insert .. select` that's completed the select part but not the insert part) so we want to give the real return in those cases. In the case of things that can be cancelled, `rx` will eventually return some cancelled-like error.

The reason for the `if !cancelled` is that the cancel future will panic if polled after successfully returning. Not sure if there's a better idiom for expressing this

Reply:
Ahhh that totally makes sense. Sorry I didn't comment sooner, but I think [`FutureExt::fuse`](https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.fuse) would fit this use case perfectly! IMO no need to go back and change it though, the `cancelled` flag works just fine